src/Type/Php/ArrayFilterFunctionReturnTypeReturnTypeExtension.php

```diff
@@ -105,12 +107,22 @@ public function removeFalsey(Type $type): Type
 				} elseif ($isFalsey->maybe()) {
 					$values[$offset] = TypeCombinator::remove($values[$offset], $falseyTypes);
 					$generalize = true;
+				} else {
+					$notEmpty = true;
```

Comment:
I think you need to use a `TrinaryLogic` for this. `array_filter(DateTime|null)` might result in an empty array, or it might not. You don't know for sure that the result will be non-empty unless you're sure of the types.